//---------------------------------------------------------------------
//---------------------------------------------------------------------
syntax = "proto3";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
import "google/protobuf/any.proto";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
option cc_enable_arenas = true;

//---------------------------------------------------------------------
//---------------------------------------------------------------------
package niPerfTest;

//---------------------------------------------------------------------
// The niPerfTest service definition.
//---------------------------------------------------------------------
service niPerfTestService {
    rpc StreamLatencyTest(stream StreamLatencyClient) returns (stream StreamLatencyServer) {};
    rpc StreamLatencyTestClient(stream StreamLatencyClient) returns (StreamLatencyServer) {};
    rpc StreamLatencyTestServer(StreamLatencyClient) returns (stream StreamLatencyServer) {};
    rpc TestWrite(TestWriteParameters) returns (TestWriteResult) {};
    rpc TestWriteContinuously(stream TestWriteParameters) returns (stream TestWriteResult) {};

    rpc Init(InitParameters) returns (InitResult) {};
    rpc ConfigureVertical(ConfigureVerticalRequest) returns (ConfigureVerticalResponse);
    rpc ConfigureHorizontalTiming(ConfigureHorizontalTimingRequest) returns (ConfigureHorizontalTimingResponse);    
    rpc InitiateAcquisition(InitiateAcquisitionRequest) returns (InitiateAcquisitionResponse);
    rpc Read(ReadParameters) returns (ReadResult) {};
    rpc ReadContinuously(ReadContinuouslyParameters) returns (stream ReadContinuouslyResult) {};
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
service MonikerService {
  rpc InitiateMonikerStream(MonikerList) returns (MonikerStreamId);
  rpc StreamReadWrite(stream MonikerWriteRequest) returns (stream MonikerReadResult) {};
  rpc StreamRead(MonikerStreamId) returns (stream MonikerReadResult) {};
  rpc StreamWrite(stream MonikerWriteRequest) returns (stream MonikerStreamId) {};
}

message Moniker {
  string ServiceLocation = 1;
  string DataSource = 2;
  string DataInstance = 3;
}

message DoubleArray {
  repeated double values = 1;
}

message IntArray {
  repeated sint32 values = 1;
}

message DataValue {
  oneof value {
    DoubleArray doubleArray = 1;
    IntArray intArray = 2;
    google.protobuf.Any Values = 3;
  }
}

message MonikerWriteRequest {
  int32 monikerStreamId = 1;
  repeated DataValue Values = 2;
}

message MonikerReadResult {
  repeated DataValue Values = 1;
}

message MonikerList {
  repeated Moniker ReadMonikers = 1;
  repeated Moniker WriteMonikers = 2;
  bool useAnyType = 3;
}

message MonikerStreamId {
  int32 streamId = 1;
}

message StreamLatencyClient {
  uint32 message = 1;
}

message StreamLatencyServer {
  uint32 message = 1;
}

message InitParameters {
  string command = 1;
  int32 id = 2;
}
    
message InitResult {
  int32 status = 1;
}

message ReadParameters {
  double timeout = 1;
  int32 numSamples = 2;
}
    
message ReadResult {
  int32 status = 1;
  repeated double samples = 2;
}

message TestWriteParameters {
  repeated double samples = 2;
}
    
message TestWriteResult {
  int32 status = 1;
}

message ReadContinuouslyParameters {
  int32 numIterations = 1;
  int32 numSamples = 2;
}
    
message ReadContinuouslyResult {
  int32 status = 1;
  repeated double wfm = 2;
}

message ConfigureHorizontalTimingRequest {
  string vi = 1;
  double min_sample_rate = 2;
  sint32 min_num_pts = 3;
  double ref_position = 4;
  sint32 num_records = 5;
  bool enforce_realtime = 6;
}

message ConfigureHorizontalTimingResponse {
  int32 status = 1;
}

enum VerticalCoupling {
  option allow_alias = true;
  VERTICAL_COUPLING_UNSPECIFIED = 0;
  VERTICAL_COUPLING_NISCOPE_VAL_AC = 0;
  VERTICAL_COUPLING_NISCOPE_VAL_DC = 1;
  VERTICAL_COUPLING_NISCOPE_VAL_GND = 2;
}

message ConfigureVerticalRequest {
  string vi = 1;
  string channel_list = 2;
  double range = 3;
  double offset = 4;
  oneof coupling_enum {
    VerticalCoupling coupling = 5;
    sint32 coupling_raw = 6;
  }
  double probe_attenuation = 7;
  bool enabled = 8;
}

message ConfigureVerticalResponse {
  int32 status = 1;
}

message InitiateAcquisitionResponse {
  enum StatusCode {
    default = 0;
    good = 1;
    bad = 2;
  }
  int32 status = 1;
}
  
message InitiateAcquisitionRequest {
  string vi = 1;
  InitiateAcquisitionResponse.StatusCode code = 2;
}
